src/mobu/services/repo.py

```diff
@@ -0,0 +1,164 @@
+        """Clone a git repo or return cached info by url + ref.
+
+        Increase the reference count for the url + ref + hash combo.
+
+        Parameters
+        ----------
+        url
+            The URL of the repo to clone
+        ref
+            The git ref to checkout after the repo is cloned
+        """
+        logger = self._logger.bind(url=url, ref=ref)
+        key = _Key(url=url, ref=ref)
+
+        async with self._lock:
+            # If the notebook repo has already been cloned, return the info
+            if info := self._cache.get(key):
+                logger.info("Notebook repo cached")
+                reference = _Reference(url=url, ref=ref, hash=info.hash)
+                count = self._references[reference]
+                count.count += 1
+                return info
+
+            # If not, clone the repo
+            logger.info("Cloning notebook repo")
+            repo_dir = TemporaryDirectory(delete=False, dir=self._dir.name)
+            with capturing_start_span(op="clone_repo"):
+                git = Git(logger=self._logger)
+                git.repo = Path(repo_dir.name)
+                await git.clone(url, repo_dir.name)
+                await git.checkout(ref)
+                repo_hash = await git.repo_hash()
+
+            # If we're in testing mode, record that we actually did a clone
+            if self._testing:
+                self._cloned.append(key)
+
+            info = ClonedRepoInfo(
+                dir=repo_dir, path=Path(repo_dir.name), hash=repo_hash
+            )
+
+            # Update the cache with the cloned repo's info
+            self._cache[key] = info
+
+            # Update the reference count
+            reference = _Reference(url=url, ref=ref, hash=info.hash)
+            self._references[reference] = _ReferenceCount(
+                count=1, dir=info.dir
+            )
+            return info
+
+    async def invalidate(self, url: str, ref: str, repo_hash: str) -> None:
+        """Invalidate a git repo in the cache by url + ref.
+
+        Decrease the url + ref + hash reference count. If it drops to zero,
+        delete the files from that clone.
+
+        Parameters
+        ----------
+        url
+            The URL of the repo to clone
+        ref
+            The git ref to checkout after the repo is cloned
+        hash
+            The hash of the cloned repo to remove
+        """
+        logger = self._logger.bind(url=url, ref=ref, hash=repo_hash)
+        key = _Key(url=url, ref=ref)
+        reference = _Reference(url=url, ref=ref, hash=repo_hash)
+
+        # This theoretically doesn't need a lock, but if we add any awaits here
+        # in the future, we'd need to add a lock, and it would be pretty easy
+        # to forget to do it.
+        async with self._lock:
+            info = self._cache.get(key)
+            if info:
+                logger.info("Invalidating repo")
+                del self._cache[key]
```

Comment:
Hm, I think this means that if any monkey shuts down, it invalidates the cache for its repository and forces a reclone even if that ends up being the same hash. I think this is fine for now because doing something more complicated requires adding the concept of checking to see if a given checkout of a repo is still current, but it might be worth a comment saying that this will result in some unnecessary clones still.